### PR TITLE
MKL support for max/avg pooling and relu

### PR DESCRIPTION
**Pull request by vivek-rane**
This has been branched off the MKL pull request for convolution so please review the last 2 commits only. Once the MKL convolution pull request is complete, we will merge the tensorflow master here and that should make things clearer.

### CODE REVIEW

**Comment by caisq** [comment]
@tensorflow-jenkins test this please

**Comment by girving** [comment]
I'm the wrong person to review this.  Cc @rmlarsen or @benoitsteiner in case they're interested.

**Comment by vivek-rane** [comment]
Please restart tests.

**Comment by andydavis1** [comment]
@tensorflow-jenkins test this please

**Comment by vivek-rane** [comment]
There's a failure with tensorflow/python:input_test that I can't reproduce on my setup. How do I access "/var/lib/jenkins/workspace/tensorflow-pull-requests-cpu/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/bazel-out/local-opt/testlogs/bazel_pip/tensorflow/python/input_test/test.log"?

File: tensorflow/core/kernels/mkl_pooling_ops_common.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  int64 pad_left;
+  int64 pad_right;
+  int64 pad_top;
+  int64 pad_bottom;
+  int pad_depth;
+
+  TensorFormat data_format;
+
+  // Updates context->status if there is an invalid input.
+  void init(OpKernelContext* context, const std::vector<int32>& ksize,
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_pooling_ops_common.h]: These should all start with upper case "Init"

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    return context->status().ok();
+  }
+};
+
+template <typename Device, typename T>
+class MklReluOp : public OpKernel {
+ public:
+  ~MklReluOp() {}
+  explicit MklReluOp(OpKernelConstruction* context) : OpKernel(context) {}
+  void Compute(OpKernelContext* context) override {
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Add a line space here

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+      mkl_params.output_shape.SetMklTensor(false);
+      AllocateOutputSetMklshape(context, 0, &out_tensor, o_shape,
+                                mkl_params.output_shape);
+      void* out_o = static_cast<void*>(out_tensor->flat<T>().data());
+      (static_cast<T*>(out_o))[0] =
+          std::max((static_cast<T*>(user_i))[0], static_cast<T>(0));
+      return;
+    }
+
+    if (input_in_mkl_format) {
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Some comments here would be great.

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+    size_t *in_sizes;
+    size_t *in_strides;
+    MklShape input_shape, grad_shape, output_shape;
+  } MklReluGradOpParams_;
+  MklReluGradOpParams_ mkl_params;
+
+  void MklPrepareReluGradInputs(OpKernelContext* context,
+                                   Tensor* mkl_tmp_grad_buf_tensor,
+                                   Tensor* mkl_tmp_input_buf_tensor) {
+  dnnPrimitive_t cv_user_to_reluB_input = nullptr,
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Put these on separate lines please.

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+      cv_user_to_reluB_grad = nullptr;
+  dnnLayout_t mkl_lt_internal_input = nullptr, mkl_lt_internal_grad = nullptr;
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Indention is strange in this function. Is this indented correctly?

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+  } else {
+    relu_res[dnnResourceDiffDst] = user_g;
+  }
+
+  if (cv_user_to_reluB_grad) dnnDelete_F32(cv_user_to_reluB_grad);
+  dnnLayoutDelete_F32(mkl_lt_internal_grad);
+  }
+
+  void MklCreateInputLayouts(OpKernelContext* context) {
+  bool grad_is_mkl = mkl_params.grad_shape.IsMklTensor();
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Again, indention is off.

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+    mkl_params.output_shape.SetMklTensor(false);
+    AllocateOutputSetMklshape(context, 0, &output,
+        g_shape, mkl_params.output_shape);
+    void* out_o = static_cast<void*>(output->flat<T>().data());
+    (static_cast<T*>(out_o))[0] =
+        (static_cast<T*>(user_g))[0] * ((static_cast<T*>(user_i))[0] > 0);
+    return;
+  }
+
+  if (grad_is_mkl || input_is_mkl) {
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Add comment as to what these two code branches do...

File: tensorflow/core/kernels/mkl_pooling_ops_common.cc [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    depth = GetTensorDim(tensor_in_shape, data_format, 'C');
+    tensor_in_cols = GetTensorDim(tensor_in_shape, data_format, 'W');
+    tensor_in_rows = GetTensorDim(tensor_in_shape, data_format, 'H');
+    tensor_in_batch = GetTensorDim(tensor_in_shape, data_format, 'N');
+
+    init(context, ksize, stride, padding, data_format);
+  }
+
+  // Initialization for MKL format
+  void MklPoolParameters::init(OpKernelContext* context,
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_pooling_ops_common.cc]: Upper case for function names: "Init". Same below...

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+    CHECK_EQ(dnnExecute_F32(mkl_prim_relu_fwd_, relu_res), E_SUCCESS);
+    Mklcleanup();
+  }
+
+ private:
+  typedef struct {
+    int in_dims;
+    size_t *in_sizes;
+    size_t *in_strides;
+    MklShape input_shape, output_shape;
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: Put these on separate lines.

File: tensorflow/core/kernels/mkl_relu_op.cc [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+  if (!dnnLayoutCompare_F32(mkl_lt_internal_grad, mkl_lt_grad_)) {
+    AllocTmpBuffer(context, mkl_tmp_grad_buf_tensor, mkl_lt_internal_grad,
+                   &relu_res[dnnResourceDiffDst]);
+    CHECK_EQ(dnnConversionCreate_F32(&cv_user_to_reluB_grad, mkl_lt_grad_,
+                                     mkl_lt_internal_grad),
+             E_SUCCESS);
+  }
+
+  if (!dnnLayoutCompare_F32(mkl_lt_internal_input, mkl_lt_input_)) {
```
Comment (andydavis1) [inline tensorflow/core/kernels/mkl_relu_op.cc]: These parameter names are very cryptic. If you can't make them more descriptive, then please add comments explaining each parameter.

File: tensorflow/core/ops/nn_ops.cc [hunk 11]
```diff
@@ -144,6 +144,45 @@ output: 4-D.  Gradients w.r.t. the input of `avg_pool`.
 
 // --------------------------------------------------------------------------
 
+#ifdef INTEL_MKL
+
+REGISTER_OP("MklAvgPool")
```
Comment (andydavis1) [inline tensorflow/core/ops/nn_ops.cc]: Please add documentation for each input/output/attr. See docstring for other ops in this file. Documentation in these sections is important. Thanks.

**Comment by vivek-rane** [comment]
Fixed some of the code review comments and rebased to the latest master to pull code from one of the other merged pull requests.

**Comment by andydavis1** [comment]
Looks good, but please address my comments before submitting.

**Comment by andydavis1** [comment]
@tensorflow-jenkins test this please

**Review by andydavis1 — APPROVED** [review]

**Comment by vivek-rane** [comment]
I think this is ready for merge - please let me know if we missed anything.

**Comment by andydavis1** [comment]
@tensorflow-jenkins test this please

**Comment by martinwicke** [comment]
@andydavis1 Are we sure we want these ops in core immediately? Could we add them to contrib, or at least expose them in the python api in contrib first?

**Comment by martinwicke** [comment]
I added [wip] to the title, remove it once the convolution change is it and this is unblocked. Thanks!

**Comment by vivek-rane** [comment]
Convolution changes have been added to the mkl-kernels pull request (https://github.com/tensorflow/tensorflow/pull/8184)

**Comment by andydavis1** [comment]
@martinwicke thanks. We are almost through all of their kernels at this point, so it would be difficult to move things to contrib first.

**Comment by martinwicke** [comment]
Ok. Maybe it's not a big problem here... how would these kernels be exposed?

**Comment by vivek-rane** [comment]
@martinwicke These kernels are applied in the graph by replacing their Eigen counterparts when MKL is enabled during configure. This option (in configure) is not enabled yet as we do not have all the pieces for MKL in the mainline yet.

**Comment by martinwicke** [comment]
jenkins, test this please.

**Pull request merged**